storage/range_gc_queue.go

```diff
@@ -0,0 +1,147 @@
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Ben Darnell
+
+package storage
+
+import (
+	"time"
+
+	"github.com/cockroachdb/cockroach/client"
+	"github.com/cockroachdb/cockroach/proto"
+	"github.com/cockroachdb/cockroach/storage/engine"
+	"github.com/cockroachdb/cockroach/util"
+	"github.com/cockroachdb/cockroach/util/log"
+)
+
+const (
+	// rangeGCQueueMaxSize is the max size of the gc queue.
+	rangeGCQueueMaxSize = 100
+
+	// rangeGCQueueTimerDuration is the duration between GCs of queued ranges.
+	rangeGCQueueTimerDuration = 10 * time.Second
+)
+
+// rangeGCQueue manages a queue of ranges to be considered for garbage
+// collections. The GC process asynchronously removes local data for
+// ranges that have been rebalanced away from this store.
+type rangeGCQueue struct {
+	*baseQueue
+	db *client.KV
+}
+
+// newRangeGCQueue returns a new instance of rangeGCQueue.
+func newRangeGCQueue(db *client.KV) *rangeGCQueue {
+	q := &rangeGCQueue{
+		db: db,
+	}
+	q.baseQueue = newBaseQueue("rangeGC", q, rangeGCQueueMaxSize)
+	return q
+}
+
+func (q *rangeGCQueue) needsLeaderLease() bool {
+	return false
+}
+
+// shouldQueue determins whether a range should be queued for GC,
+// and if so at what priority. Currently all inactive ranges are
+// considered for possible GC at equal priority.
+func (q *rangeGCQueue) shouldQueue(now proto.Timestamp, rng *Range) (bool, float64) {
+	if _, expired := rng.HasLeaderLease(now); !expired {
+		// If anyone holds a non-expired lease and we know about it, we
+		// have recently been an active member of the range.  This is not
+		// a guarantee that the range is not garbage (we could have just
+		// been removed), but leader leases are short enough that it
+		// doesn't hurt to wait for them to expire, and this gives us a
+		// cheap heuristic to avoid unnecessary queries.
+		return false, 0
+	}
+
+	// Otherwise, continue to processing.
+	return true, 0
+}
+
+// process performs a consistent lookup on the range descriptor to see if we are
+// still a member of the range.
+func (q *rangeGCQueue) process(now proto.Timestamp, rng *Range) error {
+	// Calls to InternalRangeLookup typically use inconsistent reads, but we
+	// want to do a consistent read here. This is important when we are
+	// considering one of the metadata ranges: we must not do an inconsistent
+	// lookup in our own copy of the range.
+	reply := proto.InternalRangeLookupResponse{}
+	err := q.db.Run(client.Call{
+		Args: &proto.InternalRangeLookupRequest{
+			RequestHeader: proto.RequestHeader{
+				// TODO(bdarnell): what is the right key to use here?
+				// InternalRangeLookup fails if given KeyMin (inconsistent
+				// reads of the first range use gossip instead of
+				// InternalRangeLookup).
+				Key: engine.RangeMetaKey(rng.Desc().StartKey.Next()),
```

Comment:
@spencerkimball Which key should I use here? KeyMin is treated as a special case in RangeMetaKey (RangeMetaKey(KeyMin) == KeyMin), but InternalRangeLookup panics if I pass it in. Is this a bug in InternalRangeLookup?